File: SweepLine/56. Merge Intervals/56. Merge Intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length == 0) return intervals;
        // generate sweep line
        int[][] sweep = new int[intervals.length*2][2];
        for (int i = 0; i < intervals.length; i++) {
            sweep[2*i][0] = intervals[i][0];
            sweep[2*i][1] = 1;
            sweep[2*i+1][0] = intervals[i][1];
            sweep[2*i+1][1] = -1;
        }
        Arrays.sort(sweep, (a, b) -> {
            if (a[0] == b[0]) {
                return b[1]-a[1];
            } else {
                return a[0]-b[0];
            }
        });
        List<int[]> list = new ArrayList<>();
        int count = 0;
        int[] temp = new int[2];
        for (int[] s : sweep) {
            if (s[1] == 1) {
                count++;
                if (count == 1) {
                    temp[0] = s[0];
                }
            } else {
                count--;
                if (count == 0) {
                    temp[1] = s[0];
                    //System.out.println("start -> " + temp[0] + " : end -> " + temp[1]);
                    list.add(new int[] {temp[0], temp[1]});
                }
            }
        }
        return list.toArray(new int[list.size()][]);
//         if (intervals == null || intervals.length == 0 || intervals[0].length == 0) return intervals;
//         Arrays.sort(intervals, (a, b) -> (a[0]-b[0]));
//         List<int[]> list = new ArrayList<>();
//         int start = intervals[0][0], end = intervals[0][1];
//         for (int i = 1; i < intervals.length; i++) {
//             if (intervals[i][0] <= end) {
//                 start = Math.min(start, intervals[i][0]);
//                 end = Math.max(end, intervals[i][1]);
//             } else {
//                 list.add(new int[] {start, end});
//                 start = intervals[i][0];
//                 end = intervals[i][1];
//             }
//         }
//         list.add(new int[] {start, end});
        
// //         int[][] rets = new int[list.size()][2];
// //         for (int i = 0;i < list.size(); i++) {
// //             rets[i] = list.get(i);
// //         }
// //         return rets;
//         return list.toArray(new int[list.size()][]);
    }
}

// s--------e
//     s'------e'
//     s2-----e2

File: SweepLine/56. Merge Intervals/56. Merge Intervals.cpp
class Solution {
    static bool cmp(pair<int, int> a, pair<int, int> b) {
        if (a.first == b.first) return a.second > b.second;
        else return a.first < b.first;
    }
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<pair<int, int>>sweep;
        for (auto x : intervals) {
            sweep.push_back({x[0], 1});
            sweep.push_back({x[1], -1});
        }
        sort(sweep.begin(), sweep.end(), cmp);
        
        int count = 0;
        vector<vector<int>>rets;
        int start, end;
        for (auto x : sweep) {
            if (x.second == 1) {
                count++;
                if (count == 1) start = x.first; 
            } else {
                count--;
                if (count == 0) {
                    end = x.first;
                    rets.push_back({start, end});
                }
            }
        }
        return rets;
    }
};

// 2-------4
//     3-------8
// 2-----------8
// time (start/end, label 1/-1)
    
// 1-------4
//         4-------8
